File: Year 2/Semester 4/Programming and Projection Enironments/Week 4/C#FourthWeek/SaptamanaAPatra/repository/TicketDBRepository.cs
using log4net;
using SaptamanaAPatra.domain;
using SaptamanaAPatra.repository.interfaces;
using SaptamanaAPatra.utils;
using System.Collections.Generic;
using System.Data.SQLite;

namespace SaptamanaAPatra.repository
{
    internal class TicketDBRepository : ITicketRepository
    {
        private static readonly ILog logger = LogManager.GetLogger("TicketDBRepository");

        private IDictionary<string, string> _props;

        public TicketDBRepository(IDictionary<string, string> props)
        {
            this._props = props;
            logger.Info("Creating TicketDBRepository");
        }

        private static Ticket<int> GetTicketFromResultSet(SQLiteDataReader resultSet)
        {
            int id = resultSet.GetInt32(0);
            int idShow = resultSet.GetInt32(1);
            string nameOfCostumer = resultSet.GetString(2);
            return new Ticket<int>(id, idShow, nameOfCostumer);
        }

        public List<Ticket<int>> GetAll()
        {
            logger.Info("finding all tickets");
            var connection = DBUtils.GetConnection(_props);
            var list = new List<Ticket<int>>();
            using (var cmd = new SQLiteCommand("select * from ticket", connection as SQLiteConnection))
            using (var reader = cmd.ExecuteReader())
                while (reader.Read())
                    list.Add(GetTicketFromResultSet(reader));
            logger.Info(list);
            return list;
        }


        public Ticket<int> FindById(int id)
        {
            logger.Info("find a ticket by id");
            var connection = DBUtils.GetConnection(_props);
            var cmd = new SQLiteCommand("select * from ticket where id = " + id,
                connection as SQLiteConnection);
            using (var reader = cmd.ExecuteReader())
            {
                if (reader.Read())
                {
                    Ticket<int> ticket = GetTicketFromResultSet(reader);
                    logger.Info(ticket);
                    return ticket;
                }
            }
            return null;
        }

        public bool Add(Ticket<int> ticket)
        {
            logger.Info("adding ticket");
            var connection = DBUtils.GetConnection(_props);
            using (var command = new SQLiteCommand
                ("insert into ticket values (@id, @id_show, @name_of_costumer)",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", null);
                command.Parameters.AddWithValue("@id_show", ticket.IdShow);
                command.Parameters.AddWithValue("@name_of_costumer", ticket.NameOfCostumer);
                command.ExecuteNonQuery();
                logger.InfoFormat("inserted successfully {0}", ticket);
                return true;
            }
        }


        public bool Delete(int id)
        {
            logger.InfoFormat("delete ticket with id = {0}", id);
            var connection = DBUtils.GetConnection(_props);
            using (var command = new SQLiteCommand("delete from ticket where id = @id",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", id);
                command.ExecuteNonQuery();
                logger.InfoFormat("deleted successfully for id {0}", id);
                return true;
            }
        }

        public bool Update(Ticket<int> ticket)
        {
            logger.Info("updating employee");
            var connection = DBUtils.GetConnection(_props);
            using (var command = new SQLiteCommand("update ticket set id_show = @id_show, " +
                "name_of_costumer = @name_of_costumer where id = @id",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id_show", ticket.IdShow);
                command.Parameters.AddWithValue("@name_of_costumer", ticket.NameOfCostumer);
                command.Parameters.AddWithValue("@id", ticket.Id);
                command.ExecuteNonQuery();
                logger.Info("updated successfully");
                return true;
            }
        }
    }
}

File: Year 2/Semester 4/Programming and Projection Environments/Week 4/JavaFourthWeek/src/main/java/repository/TicketDBRepository.java
package repository;

import domain.Ticket;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class TicketDBRepository extends AbstractRepository<Integer, Ticket> {

    public TicketDBRepository(Properties props) {
        super(props);
    }

    private Ticket getTicketFromResultSet(ResultSet resultSet) throws SQLException {
        int id = resultSet.getInt("id");
        int idShow = resultSet.getInt("id_show");
        String nameOfCostumer = resultSet.getString("name_of_costumer");
        return new Ticket(id, idShow, nameOfCostumer);
    }

    @Override
    public Ticket findById(Integer id) {
        logger.traceEntry("find a ticket by id");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "select * from ticket where id = ?")) {
            ps.setInt(1, id);
            try (ResultSet resultSet = ps.executeQuery()) {
                if (resultSet.next()) {
                    Ticket ticket = getTicketFromResultSet(resultSet);
                    logger.traceExit(ticket);
                    return ticket;
                }
            }
        } catch (SQLException exception) {
            logger.error("ERROR for findById in TicketDBRepository: " + exception);
        }
        return null;
    }

    @Override
    public Iterable<Ticket> getAll() {
        logger.traceEntry("finding all tickets");
        Connection connection = dbUtils.getConnection();
        List<Ticket> tickets = new ArrayList<>();
        try (PreparedStatement ps = connection.prepareStatement("select * from ticket")) {
            try (ResultSet resultSet = ps.executeQuery()) {
                while (resultSet.next()) {
                    tickets.add(getTicketFromResultSet(resultSet));
                }
            }
        } catch (SQLException exception) {
            logger.error("ERROR for getAll in TicketDBRepository: " + exception);
        }
        logger.traceExit(tickets);
        return tickets;
    }

    @Override
    public boolean add(Ticket ticket) {
        logger.traceEntry("Buying ticket {}", ticket);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "insert into ticket (id_show, name_of_costumer) values (?, ?)")) {
            ps.setInt(1, ticket.getIdShow());
            ps.setString(2, ticket.getNameOfCostumer());
            int result = ps.executeUpdate();
            logger.trace("saved {} instance", result);
        } catch (SQLException exception) {
            logger.error("ERROR for insert in TicketDBRepository: " + exception);
            return false;
        }
        logger.traceExit("inserted successfully");
        return true;
    }

    @Override
    public boolean delete(Integer id) {
        logger.traceEntry("delete ticket");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("delete from ticket where id=?")) {
            ps.setInt(1, id);
            int result = ps.executeUpdate();
            logger.trace("deleted {} instance", result);

        } catch (SQLException exception) {
            logger.error("ERROR for delete in TicketDBRepository: " + exception);
            return false;
        }
        logger.traceExit("deleted successfully");
        return true;
    }

    @Override
    public boolean update(Ticket ticket) {
        logger.traceEntry("updating ticket");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "update ticket set id_show = ?, name_of_costumer = ?" +
                " where id = ?")) {
            ps.setInt(1, ticket.getIdShow());
            ps.setString(2, ticket.getNameOfCostumer());
            ps.setInt(3, ticket.getId());
            int result = ps.executeUpdate();
            logger.trace("updated {} instance", result);
        } catch (SQLException exception) {
            logger.error("ERROR for update in TicketDBRepository: " + exception);
            return false;
        }
        logger.traceExit("updated successfully");
        return true;
    }
}